### DISCUSSION TITLE
Layout 컴포넌트의 개발 방식

### DISCUSSION THREAD

**Discussion by mina-gwak** [Co-Studo/Co-Studo-front#16] [General]
기능이 존재하지 않는, 즉 **레이아웃을 잡는 용도로 쓰이는 컴포넌트를 어떤 방식으로 개발**할지 이야기해보고 통일하면 좋을 것 같아요!

```tsx
const Row = ({ title, description, left, right }) => (
  <S.Container>
    <S.TitleWrapper>
      <S.Left>{left}</S.Left>
      <S.Title>{title}</S.Title>
      <S.Right>{right}</S.Right>
    </S.TitleWrapper>
    <S.Description>{description}</S.Description>
  </S.Container>
);
```

위의 코드는 제가 생각한 Row 컴포넌트의 구조이고, 고민해보았던 부분을 정리하면 다음과 같습니다.

- 다른 Headless 컴포넌트들을 만든 방식처럼 개별 컴포넌트를 합성해서 사용하도록 하는 방식은 레이아웃을 잡는다는 용도에 적합하지 않은 것 같다. (정해진 틀이 있어야 하는데 유연하게 변경될 수 있게 되므로)
- 그래서 이전까지 컴포넌트를 개발하던 방식대로 하고 (공통적인 스타일 포함), 변경이 되는 요소들은 props로 받도록 해야겠다고 생각
- 만든 후 해당 컴포넌트가 쓰이게 될만한 곳들을 디자인에서 확인해보니, 공통적으로 쓰여서 레이아웃 컴포넌트에 지정해주면 된다고 생각했던 폰트 사이즈나 굵기, 여백 등 스타일이 달라지는 부분이 존재 => 각 개별 스타일까지 설정할 수 있도록 옵션을 받기엔,,, 뭔가 아닌 것 같다 => 변경이 되는 스타일은 props로 받는 요소에서 설정하도록 하고 변경이 되지 않는 부분만 이 컴포넌트에서 설정하자니 어떤 게 변경이 되고 변경이 되지 않는지 매번 확인이 필요하고 각 컴포넌트마다 달라질 수 있는 부분이기 때문에 좀 번거로울 것 같다 => 아직 마땅한 해결 방법 생각 못함

그래서 어떻게 개발할지 의견을 나눠보면 좋을 것 같습니다! 

@jindonyy @hemudi @healtheloper

**Comment by hemudi** [comment]
좋은 논의 주제인거 같아요! 저도 마침 고민이 되던 부분이었어요! ^_^)b

확실히 모든 스타일들을 설정할 수 있게 옵션을 받는건 props의 양이 너무 많아지고, 과한 범용성을 가진 컴포넌트가 될 수 있으니
컴포넌트에서 커스텀 가능하도록 허용하는 속성을 같이 논의하여 한정지어 놓는건 어떨까요?
그렇게해서 커스텀 가능한 범위를 벗어난 비슷한 형태의 컴포넌트는 그냥 별도로 분리하여 다른 컴포넌트로 만드는 식으로요.
어차피 이러한 컴포넌트는 저희 프로젝트에서만 쓰일거니까 과하게 커스텀 가능하지 않아도 괜찮을거 같아요.

예를 들어
row 같은 이러한 컴포넌트들은 레이아웃을 잡는게 목적이니 컴포넌트의 여백이나 크기 같은건 고정으로 하고
폰트, 글자 색 등 꾸미기 요소만 props로 받는 식...?

아니면 반대로 커스텀 불가능한 속성을 정해놓는게 빠를 수도 있겠네요!

**Reply by mina-gwak** [reply]
> 컴포넌트에서 커스텀 가능하도록 허용하는 속성을 같이 논의하여 한정지어 놓는건 어떨까요?

좋은 것 같아요! 말씀하신대로 저희 컴포넌트에서 쓰이는 것들이니 저희끼리의 룰을 정해두고 하면 번거로운 부분도 좀 줄어들고 커스텀도 어느정도 가능해지니 괜찮은 방안인 것 같습니다.
다만 스타일링을 위한 속성이 아닌 여백 같은 경우는 스타일이 거의 비슷해도 조금씩 다른 경우가 있어서 이런 부분을 어떻게 처리할지 추가로 논의해서 정해두면 되지 않을까 싶어요!

**Reply by healtheloper** [reply]
Row 컴포넌트에 대한 grid column gap 을 정해두고 쓰면 되지 않을까요?!

**Comment by healtheloper** [comment]
노션에 올린 레이아웃 예시중에 right 를 잘못잡은 게 하나 있는데 (즐겨찾기는 card 에 들어가는게 맞는 것 같아요)
그래서 title wrapper 에 left, title, right 가 들어간 걸로 보이네용

![image](https://user-images.githubusercontent.com/58503584/191663622-4dca5997-4aaf-4527-8855-277a37ec3c6a.png)

> 레이아웃을 잡는다는 용도에 적합하지 않은 것 같다. (정해진 틀이 있어야 하는데 유연하게 변경될 수 있게 되므로)

정해진 틀이라는게 제가 생각했던 건 **content width 가 있으면 left, contents(title + description), right 가 grid 로 적절히 나눠질 걸 생각했던 것 같아요**, 그래서 어떤 컴포넌트를 넘겨주던 그 크기에 맞게 width 가 잡히도록?

<img width="287" alt="스크린샷 2022-09-22 오후 2 18 07" src="https://user-images.githubusercontent.com/58503584/191664142-3699394b-d4a0-413f-b655-51bb10f4ee36.png">

만약 title, description 을 둘 다 받는다면 아래 형태가 되고

<img width="313" alt="스크린샷 2022-09-22 오후 2 18 54" src="https://user-images.githubusercontent.com/58503584/191664412-617cbe80-f232-48d0-9de8-5b2792368b37.png">


만약 left 가 없다면 이런 형태가 될 거라고 생각하고 grid container 로 생각했던 것 같아요.

<img width="341" alt="스크린샷 2022-09-22 오후 2 19 28" src="https://user-images.githubusercontent.com/58503584/191664425-13fed6f4-bdfb-4008-b81f-a14c908833bb.png">

결국 어떤 컴포넌트를 받던지 저 레이아웃 안에 넣는다는 건데, 그렇다면 Component 자체를 prop 으로 받는 형식이 낫지 않을까요?

**Reply by mina-gwak** [reply]
> 정해진 틀이라는게 제가 생각했던 건 content width 가 있으면 left, contents(title + description), right 가 grid 로 적절히 나눠질 걸 생각했던 것 같아요

Grid를 사용하려고 하진 않았지만 저도 파크가 말씀하신 것과 비슷하게 생각했어요! 위의 Row 컴포넌트가 받고 있는 props들은 컴포넌트를 받는다고 생각하고 만들었어요. 근데 이제 레이아웃을 잡는 컴포넌트니까 props로 받게 되는 컴포넌트들을 위치시키기 위해서 `S.Left` 와 같은 컨테이너로 감싸주고 포지셔닝해서 위치를 잡으려고 했었습니다. 
<br/>

<img width="203" alt="스크린샷 2022-09-19 오후 4 21 56" src="https://user-images.githubusercontent.com/62706988/191692230-22b51b0b-b189-45a1-90e0-e61bbc85c956.png">

그리고 즐겨찾기가 카드에 들어가는 게 맞는 것 같다고 하셨는데, 저도 동의합니다! 그렇다고 한다면 알림의 X 버튼도 Row가 아니라 알림 컴포넌트에 포함되어야 할 것 같아요. 그러면 Right 부분은 컴포넌트에서 빠져도 되지 않을까 싶네요..!

**Reply by jindonyy** [reply]
right도 디자이너랑 상의해보고 정해야할거 같아여 밑에 설명 글이 길어질 때 x 버튼 밑에 까지 글자가 갈거냐 안갈거냐를?
글자가 x 밑에까지 가면 right가 필요없을거 같고 끝까지 안간다면 right가 필요할거 같다는 생각이 드네용!

**Reply by healtheloper** [reply]
알림 컴포넌트는 X 버튼을 right 에 두고, 알림 컴포넌트의 right component 의 handler 를 알림 컴포넌트 내부에서 선언하는 게 좋지않을까요?

아래와 같은 형태로 만들고, Container 에 grid 속성을 주면 S.Left 같은 컨테이너가 필요할까요?

```jsx
<Container css={{display:grid ... }}>
  {left}
  {contents}
  {right}
</Container>
```

**Reply by healtheloper** [reply]
> right도 디자이너랑 상의해보고 정해야할거 같아여 밑에 설명 글이 길어질 때 x 버튼 밑에 까지 글자가 갈거냐 안갈거냐를? 글자가 x 밑에까지 가면 right가 필요없을거 같고 끝까지 안간다면 right가 필요할거 같다는 생각이 드네용!

@jindonyy  제가 디자이너는 아니지만 안가는게 맞는 것 같습니다 ㅋㅋㅋ

**Reply by jindonyy** [reply]
> @jindonyy 제가 디자이너는 아니지만 안가는게 맞는 것 같습니다 ㅋㅋㅋ

저도 그렇게 생각합니다 ㅎㅅㅎ

**Reply by mina-gwak** [reply]
@healtheloper 아하 디자인을 지금 다시 보니 빨간 동그라미가 알림 컴포넌트 padding 내에 들어가있네요..! 제가 잘못 본 것 같아요 ㅋㅋㅠㅠ 저는 바깥 부분에 있는 줄 알고 absolute를 사용해서 배치하려고 했던 건데 말씀하신 것처럼 내부 레이아웃에 포함시켜서 만들면 될 것 같습니당.

그리고 알림 컴포넌트는 right에 X 버튼을 포함하는 게 좋을 것 같다고 하셨는데 저는 즐겨찾기 버튼이나 X 버튼이나 크게 다를 게 없다는 생각이 들거든요. 왜 둘을 다르게 생각하셨는지 알려주실 수 있을까요?

**Reply by healtheloper** [reply]
@mina-gwak 즐겨찾기 버튼은 카드 정보 레이아웃에 포함될 버튼이라고 생각이 들고, X 버튼은 알림 컴포넌트에 포함될 애라고 생각을 했어요.
그리고 알림 컴포넌트는 만드려고하는 Row 컴포넌트로 만들 수 있다고 생각이 들었습니답

<img width="668" alt="스크린샷 2022-09-22 오후 5 52 09" src="https://user-images.githubusercontent.com/58503584/191702984-946b6da0-955b-4cd2-a2ec-802dd640ef5c.png">

**Reply by mina-gwak** [reply]
@healtheloper 대충 어떤 느낌으로 말씀하신 건지 알 것 같긴 한데 이건 약간 관점 차이일 것 같네요..! 저는 Row 컴포넌트에 Right을 포함하는 방식으로 만든다면 알림 컴포넌트에 X 버튼을 넣는 것처럼 즐겨찾기 버튼도 동일한 방식을 사용해 카드 정보 레이아웃에 해당 버튼이 들어갈 레이아웃을 굳이 추가하지 않고 Row 컴포넌트만으로도 해결할 수 있는 부분이라는 생각이 들어요!

**Reply by healtheloper** [reply]
@mina-gwak 그렇네용 저는 약간 디자인 위치상으로 애매한 위치에 있는 것 같아서 별도로 레이아웃을 짜는게 맞지 않을까 했는데 굳이 라는 느낌이 또 들긴 하네요 ㅋㅋ

**Reply by hemudi** [reply]
확실히 즐겨찾기가 어디에 포함될지는 관점의 차이인거 같은데ㅎㅎ
제 관점에서는 제이미와 같이 알림 컴포넌트에서 X가 Row에 포함된다면 즐겨찾기도 Row에 포함되게 구현하는게 좀 더 자연스러운 느낌이 드는거 같네요!

**Comment by jindonyy** [comment]
또 꼴등으로 다네영 ㅎㅎ
저는 두가지 방식이 생각나네여!

1. 컴포넌트 자체를 받아오기
아까 말했던 방식인데 위에 파크도 말했듯이 컴포넌트 자체를 prop으로 받는거예요.
저희가 걱정하는게 폰트 변경과 title과 description의 간격이자나요. 그게 설정되있는 컴포넌트를 prop으로 받으면 row 컴포넌트는 둘의 레이아웃(위 아래로 배치되있는) css만 신경써주면 되니까유? 사실 지금 row 컴포넌트의 위 아래 배치는 css를 따로 안줘도 되는거라 필요가 없어보이지만 예를 들어 가로로 3개 나열되는 레이아웃이 중복되면 display: flex; justify-content: space-between 이런 식으로? 그리고 그 안에 들어가는 3개의 요소들을 자기 스타일이 있는 컴포넌트 prop을 받는거얘요. 그럼 안에 요소를 신경 안쓰고 전체 레이아웃만 신경써서 prop이 줄어들 수 있다고 생각합니다!

2. 디자인 시스템에 맞게 설정
지금 딱 저희 와이어프레임에 맞게 디자인만 나온 상황이라 디자인 시스템이 없는거 같아요. 실제 플젝할 때 저희 코쿼에서 했던거처럼 폰트 사이즈와 굵기를 뭐뭐 쓸거다 정해지기 때문에 디자인 시스템에 맞게 그냥 폰트를 설정해도 되지 않을까 싶어요. title로 30, 24, 20을 쓸거다 했을 때 30일 때는 설명글이 22, 24일 때는 18 이런식으로 정해져있지 title이 30인데 설명글이 갑자기 18이 되는 경우가 잘 없고 있더라도 디자이너의 의도가 뚜렷하지 않다면 변경 요청을 해야한다고 생각해요. 간격도 폰트 크기에 맞게 간격이 정해지는 경우가 대부분이라서 디자인 시스템을 정하고 거기에 맞게 폰트 사이즈 타입을 프롭으로 받으면 어떨까 싶네요!

**Reply by mina-gwak** [reply]
1. 아하 이해했습니다! 저는 컴포넌트 내부의 세부적인 스타일까지 설정해줘야한다고 계속 생각하고 있었는데 도니가 말씀하신대로 레이아웃 컴포넌트가 정말 전체적인 틀만 잡는 느낌이라고 한다면 고민해야 할 부분이 훨씬 줄어들겠네요! 세부적인 스타일은 prop으로 받는 컴포넌트에서 설정해주면 되니까요! 저는 개인적으로 도니가 말씀하신 방식이 더 나은 것 같은데 다른 분들 의견도 들어보고 선택하면 될 것 같아요.

2. 말씀하신 것처럼 다른 분들과 논의해서 디자인 시스템을 정하면 좋을 것 같네요! 위에서 말씀하신 Right 같은 것도 포함해서 저희끼리 이야기 나눠보고 수정사항이 있다면 정리해서 제가 디자이너에게 전달하던지 아님 혹시 괜찮다고 하시면 시간을 맞춰서 함께 이야기해보던지 하면 좋을 것 같아요!

**Reply by healtheloper** [reply]
가운데 컨텐츠의 CSS 관련 변동사항이 고민이라면 Compound 식으로 여러줄을 받을 수 있는 컴포넌트 같은 걸 추가로 만들어 보는건 어떤가요?
Text 라는 컴포넌트같은걸 만들어서 저희 프로젝트에 맞는 다양한 텍스트들을 prop 으로 생성할 수 있게하고, (mui 의 [Typhography](https://mui.com/material-ui/api/typography/) 같은?)
그리고 Text Component 를 여러 줄을 받는 형태로?

ex)
```jsx

Row.한줄 = (string, props) => <Text {...props}>{string}</Text>

Row.두줄 = (top, topProps, bottom, bottomProps) => 
  (<div css={{...}}>
    <Text {...topProps}>{top}</Text>
    <Text {...bottomProps}>{bottom}</Text>
  </div>)
```

**Comment by mina-gwak** [comment]
여러 의견이 나와서 어떻게 할지는 내일 함께 이야기하면서 정하는 게 빠를 것 같아요! 많은 의견 감사합니당 👍
추가로 말씀해주실 게 있다면 코멘트 달아주세요!

@jindonyy @hemudi @healtheloper
